### PR TITLE
Fix H2 database case-sensitive identifier issues + upgrades JUnit tests cases to version 5

### LINKED ISSUE #2385: Liquibase 4.7.0 upgrade causing issue with table name
I have a hobby project running on Spring Boot 2.7 snapshot which today upgraded from Liquibase 4.6.2 to 4.7.0.

With Liquibase 4.6.2 and H2 datasource the following changelog

```
databaseChangeLog:
  - changeSet:
      id: 1
      author: me
      changes:
        - createTable:
            tableName: user
            columns:
              - column:
                  name: id
                  type: int
                  autoIncrement: true
                  constraints:
                    nullable: false
                    primaryKey: true
                    primaryKeyName: user_pk
              - ..
```

will create UPPERCASE tables in H2:

image

In fact I never really paid attention to the uppercase/lowercase table/column name in the H2 console despite having them lowercase in the changelog.

Also a changelog like this (lowercase names) works:

```
        - insert:
            tableName: user_role
            columns:
              - column:
                  name: user_id
                  valueComputed: select id from user where username = 'foo_bar'
              - column:
                  name: role_name
                  value: ROLE_USER
```

Doing some manual queries in the H2 console work for these:

- `SELECT * FROM USER`
- `SELECT * FROM user`
- `SELECT * FROM "USER"`

but not for `SELECT * FROM "user"`

Now with upgrade to Liquibase 4.7.0 I see other behaviour with above setup.

The application won't start as the changelog with the `insert` fails now:

```
Reason: liquibase.exception.DatabaseException: Table "USER" not found; SQL statement:
INSERT INTO PUBLIC.user_role (user_id, role_name) VALUES (select id from user where username = 'foo_bar', 'ROLE_USER') [42102-200] [Failed SQL: (42102) INSERT INTO PUBLIC.user_role (user_id, role_name) VALUES (select id from user where username = 'foo_bar, 'ROLE_USER')]
```

Although "USER" (uppercase) is not used like that in the insert it still fails.

Now I probably should have not used the table name `user` as `USER` is a reserved word, so I can only blame myself ;-) but this might impact others as well...

In the Liquibase 4.7.0 [release notes](https://docs.liquibase.com/release-notes/home.html) I don't see anything noteworthy that changed that could explain this behaviour.

Is this unexpected and a bug or expected behaviour?

### LINKED ISSUE #3289: Liquibase 4.8 breaks H2 1.x compatibility
Since upgrading to liquibase 4.8.0 our application fails to create the schema and load data successfully regarding a table (named 'SETTINGS') which has a column named "Value".

Before with liquibase 4.5 "Value" was not marked as a reserved word and during creation of the table the column was not quoted so it'd be created by default in Uppercase.

Since liquibase 4.8.0 the column name is seen as a reserved word and the column name is quoted during creation. So not this column gets created in actual mixed-case "Value" (instead of VALUE).

Besides ending up with a table with one column 'uppercased' and the other 'case preserved' (NAME,Value) , data loading and our application code fails:
-  loadData fails from csv file with column names in uppercase:
```
"NAME","VALUE"
"passwords.enableComplexity","Medium"
"passwords.enableAge","false"
"passwords.enableHistory","true"
```

[ so column names in loadData are not checked for reserved words and quote automatically as for other statements...? ]

Our application code fails doing a select Value from SETTINGS where Name='xxx' as the application code is not designed to quote column names as liquibase 4.8 now started to do for the "Value" column ...

Why not keep a set of V1 keywords (keywordsV1 ) and keywords (V2) and check for the major version number as is done elsewhere in the H2Database class?

e.g. here:

```
@Override
    protected String getAutoIncrementClause() {
        try {
            if (getDatabaseMajorVersion() == 1) {
```

### PR DESCRIPTION
**Pull request by mches**
## Impact

- [x] Bug fix (non-breaking change which fixes expected existing functionality)
- [ ] Enhancement/New feature (adds functionality without impacting existing logic)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
 
## Description

On H2,
1. Restore compatiblity with Liquibase 4.6.2 with respect to reserved words for H2 1.x.
2. Fold identifiers that must be quoted due to being an illegal identifier or reserved word to uppercase so that all names are in a consistent case.
3. Allow override of the defaults by way of `objectQuotingStrategy` on the change log or change set, `liquibase.preserveSchemaCase` configuration property, `modifySql`, or extension.

Fixes #2385
Fixes #3289

## Things to be aware of

* Imported the JUnit 5 BOM for better coordination of dependency versions, as a separate commit.
* Corrected whitespace errors, e.g. tab indentation, in JUnit 4 database tests a separate commit.
   * IntelliJ IDEA ignores this commit for Git Blame by default.
   * It's also possible to ignore this commit in other tools by adding it to `.git-blame-ignore-revs`.
* Converted JUnit 4 database tests to JUnit 5, as a separate commit.
* Upgraded Groovy from 2.4 to 2.5, as a separate commit.
   * This allows use of the `tap` closure.
* Configured Maven to also execute tests named `*Spec`, as a seprate commit.
   * This allows `H2DatabaseTest` can `H2DatabaseSpec` to coexist, so that the existing test doesn't have to be renamed or converted, along with it's class hierachy, to free up the name.
* Established baseline specs for H2 and all other core databases, as a separate commit.
   * Tests were passing after each commit.
   * This makes it easier to see where potential bugs exist or when bugs are fixed or behavior changed.
## Things to worry about

Some existing users' change sets or databases may require modification if they're targeting a long-lived embedded H2 database and have chosen names that require implicit quoting such as reserved words, e.g. `user`, `group`, `groups`, or `value`, or names with spaces.

## Additional Context

Spring Boot users upgrading to 2.7.x and using jOOQ Open Source Edition often have to downgrade H2 to 1.4.x. See https://github.com/spring-projects/spring-boot/issues/31168, https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-2.7-Release-Notes#jooq.

If desired, I can create separate PRs for 
1. Testing improvements
2.  H2 1.x reserved words
3.  Identifier case folding or preservation with respect to `objectQuotingStrategy`

### CODE REVIEW

**Comment by nvoxland** [comment]
Initial review:

Thanks for the fix, and the other general cleanup in there to, @mches . I appreciate the separate commits for each step you did.

The conversion to JUnit 5 tests is nice. We've generally been converting from junit 4 tests to spock tests, but that takes time for each existing test when there is just a small change you'd like to make. Having them be the nicer junit 5 will help the times when it's not a good time to redo the whole test, but still want to take advantage of nicer syntax.

I also like the separate reserved word list for v1 vs v2. That helps us support both versions, which we need to continue doing.

In your last commit, you override `quoteObject()` which I don't think is the right spot for the fix. That method is generally used as "quote the object" where the caller is expecting it to be quoted. I see it's missing javadoc specifying that, though... So it seems like the "should we quote?" logic needs to be pushed upstream to determine whether the quoteObject method should be called or not? I didn't step through where you are seeing the problem to know where that is not working, but is there a better place?

Things to be aware of:
- Nothing

**Comment by mches** [comment]
Thank you for your time and feedback @nvoxland. Of course you're right about the purpose of each of the methods. I took that to heart and came up with a better organization of the logic. Hope you like it.

**Review by filipelautert — APPROVED** [review]
This PR is restoring compatibility with H2 1.x reserved words, and as an "extra" it also upgrades JUnit tests cases to version 5.
Thanks for those fix and improvements @mches  . Also, I added some "merge branch master" commits - please feel free to rebase to get rid of them .

**Comment by mches** [comment]
@filipelautert One a separate note, after having it sit for awhile, I'm unsure whether `ObjectQuotingStrategy.LEGACY` and `ObjectQuotingStrategy.QUOTE_RESERVED_WORDS_ONLY` should be treated differently on a database other than PostgresSQL, based on the Javadoc.

```java
package liquibase.database;

/**
 * Strategy regards quoting object names e.g. table, column, index names etc.
 */
public enum ObjectQuotingStrategy {
    LEGACY, // same behavior as in liquibase 2.0, currently only postgresql has quoting handling
    QUOTE_ALL_OBJECTS, // every object gets quoted e.g. person becomes "person"
    QUOTE_ONLY_RESERVED_WORDS // quote reserved keywords and invalid column names

}
```

For the most part `QUOTE_ALL_OBJECTS` preserves object case on all databases. Does `QUOTE_RESERVED_WORDS_ONLY` imply object case of illegal identifiers and reserved words should be preserved, or just non-legacy handling of object names on PostgresSQL.

**Comment by rberezen** [comment]
@mches, I greatly appreciate these code modifications you've made! Could you kindly help resolve conflicts in the branch?

**Comment by mches** [comment]
> @mches, I greatly appreciate these code modifications you've made! Could you kindly help resolve conflicts in the branch?

@rberezen Certainly. Done! ✨

File: liquibase-standard/src/test/java/liquibase/database/core/OfflineDatabaseTest.java [hunk 1]
```diff
@@ -10,44 +10,47 @@
...
+import static org.junit.jupiter.api.Assertions.assertTrue;
 
 public class OfflineDatabaseTest {
 
-	private Database createOfflineDatabase(String url) throws Exception {
-		DatabaseConnection databaseConnection = new OfflineConnection(url, new JUnitResourceAccessor());
-		return DatabaseFactory.getInstance().openDatabase(url, null, null, null, null);
-	}
+    private Database createOfflineDatabase(String url) throws Exception {
+        DatabaseConnection databaseConnection = new OfflineConnection(url, new JUnitResourceAccessor());
```
[the remark this answers is not in the document]
Comment (rberezen) [reply liquibase-standard/src/test/java/liquibase/database/core/OfflineDatabaseTest.java:22]: @mches, would you mind reviewing this issue and a snyk security check failure?
https://app.snyk.io/org/datical/pr-checks/4a597537-a529-49b2-98bc-59e6ee008675?fromGitHubAuth=true

[the remark this answers is not in the document]
Comment (mches) [reply liquibase-standard/src/test/java/liquibase/database/core/OfflineDatabaseTest.java:22]: @rberezen I've fixed the CodeQL issue. I can look at the Snyk security check failure if you fill me in on it or grant me access to see for myself.

[the remark this answers is not in the document]
Comment (rberezen) [reply liquibase-standard/src/test/java/liquibase/database/core/OfflineDatabaseTest.java:22]: @mches Unfortunately, I cannot grant you access. However, I have shared an image depicting the error.
image

[the remark this answers is not in the document]
Comment (mches) [reply liquibase-standard/src/test/java/liquibase/database/core/OfflineDatabaseTest.java:22]: @rberezen Thank you. Fixed.

**Review by rberezen — APPROVED** [review]
Thank you very much @mches for your contribution!

**Comment by filipelautert** [comment]
@suryaaki2 could you review this one too?

**Review by karepu — APPROVED** [review]

**Review by suryaaki2 — APPROVED** [review]

**Comment by mches** [comment]
Resolved the conflict in `liquibase-extension-testing/src/main/groovy/liquibase/extension/testing/command/CommandTests.groovy`

**Comment by mches** [comment]
Rebased onto latest `master`

**Comment by filipelautert** [comment]
Thanks for the rebase! I'll get it merged today.

**Comment by nealeu** [comment]
How do we take this change out for a test ride before the release?  Should I be able to get it from Github packages without an access token for your repo?

**Comment by filipelautert** [comment]
Hey @nealeu , are you able to download it from here -> https://github.com/liquibase/liquibase/suites/15586269682/artifacts/891121515 ?

**Comment by shsmysore** [comment]
My H2 database has a table with one of the column name - `key`. (lower case reserved keyword)
How can I stop liquibase from automatically converting it to uppercase?

I tried adding below line in the changeset, but its not working.
```
<property name="preserveSchemaCase" value="true"/>
```

**Comment by mches** [comment]
> My H2 database has a table with one of the column name - `key`. (lower case reserved keyword) How can I stop liquibase from automatically converting it to uppercase?
> 
> I tried adding below line in the changeset, but its not working.
> 
> ```
> <property name="preserveSchemaCase" value="true"/>
> ```

Liquibase is uppercasing the reserved word `key` before quoting it in order to have it be treated case-insensitively. If you still want the `key` column to be lower-case, and therefore case-sensitive, then use something like this.

```xml
  <changeSet …>
    …
    <modifySql>
      <regExpReplace replace="&quot;KEY&quot;" with="&quot;key&quot;"/>
    </modifySql>
  </changeSet>
```

If you want to preserve the case of every identifier, then use someting like this.

```xml
  <changeSet … objectQuotingStrategy="QUOTE_ALL_OBJECTS">
    …
  </changeSet>
```

**Comment by shrralis** [comment]
@mches thanks for providing solutions!
The option with `modifySql` works great for me!
Anyway, it looks like it doesn't work for `loadData`, right?

**Comment by mches** [comment]
> @mches thanks for providing solutions! The option with `modifySql` works great for me! Anyway, it looks like it doesn't work for `loadData`, right?

@shrralis Yes, whether a bug or undocumented limitation, Liquibase doesn't seem to apply SQL visitors like `modifySql` to prepared statements like it does to raw SQL statements. The `loadData` change often ends up implicitly using prepared statements.

Knowing that limitiation, you may be able to work around it with something like this, but in some cases prepared statements are required.

```xml
  <changeSet …>
    <loadData … usePreparedStatements="false"/>
    <modifySql>
      …
    </modifySql>
  </changeSet>
```

You could also use object quoting strategy, but then also uppercase the CSV file's headers other than the illegal identifiers.

```xml
  <changeSet … objectQuotingStrategy="QUOTE_ALL_OBJECTS">
    <loadData …/>
  </changeSet>
```

**Comment by sixcorners** [comment]
@mches If uppercasing the quoted reserved word is done to maintain case insensitivity how do you access the reserved word case insensitively? Isn't that not really possible?

**Pull request merged**